docs/building-apps/interact-with-canisters/agents/rust-agent.mdx

```diff
@@ -30,34 +30,167 @@ Before using the agent in your project, it must be initialized using the `Agent:
+        .expect("failed to build HTTP client");
+
+    let route_provider: DynamicRouteProvider<LatencyRoutingSnapshot> =
+        DynamicRouteProviderBuilder::new(routing_strategy, seed_nodes, Arc::new(client))
+            // Set how often to fetch the latest API boundary node topology
+            .with_fetch_period(Duration::from_secs(10))
+            // Set how often to perform health checks on the API boundary nodes
+            .with_check_period(Duration::from_secs(2))
+            // Or optionally provide a custom node health checker implementation
+            // .with_checker(custom_checker)
+            // Or optionally provide a custom topology fetcher implementation
+            // .with_fetcher(custom_fetcher)
+            .build()
+            .await;
+
+    let agent = Agent::builder()
+        .with_arc_route_provider(Arc::new(route_provider))
+        .build()
+        .expect("failed to build ic-agent");
+
+    // ... use the agent for making API calls
+}
+```
+NOTE: To make use of custom dynamic routing, you must enable the currently internal feature by using:
+
+```
+cargo run --features _internal_dynamic-routing
+```
+
+##### How Dynamic Routing Works
+
+At the core of dynamic routing in the agent is the `RouteProvider trait`, which dynamically selects a routing URL for each request by invoking its `route()` method:
 
 ```rust
-let agent = Agent::builder()
-    .with_url(URL)
-    .with_background_dynamic_routing()
-    .await
-    .build()
-    .expect("failed to build ic-agent");
+pub trait RouteProvider {
+    /// Generates the next routing URL based on the internal routing logic
+    fn route(&self) -> Result<Url, AgentError>;
+    /// Generates up to `n` different routing URLs in order of priority.
+    fn n_ordered_routes(&self, n: usize) -> Result<Vec<Url>, AgentError>;
+    ...
+}
 ```
 
+In the example above, we instantiate a `DynamicRouteProvider`, an implementation of `RouteProvider`, and inject it into the agent to handle dynamic request routing. 
+The `DynamicRouteProvider` itself is generic and extensible, allowing you to define custom routing strategies. 
+
+Two built-in strategies are available:
+- Round-robin - distributes requests evenly across all healthy nodes
+- Latency-based weighted round-robin - prioritizes healthy nodes with the lowest observed latency
+In our example, we used the latency-based strategy to prioritize faster nodes, limiting the selection to the top three.
+
+Under the hood, `DynamicRouteProvider` builds and maintains a dynamic routing table, which is continuously updated by running two background services:
+- Discovery service - periodically fetches the latest API boundary node topology
+- Health check service - periodically checks nodes health by pinging it
+This design enables the agent to adapt to changes in node availability and network topology, resulting in a resilient and performant routing.
+
+For more detailed information take a look at the [documentation]("https://TODO").
```

Comment:
Just putting it here, so we don't forget to fix the link.

Reply:
I think publishing the documentation in `ic-agent` will take some days. How about i remove the link for now?

Reply:
That's also good.